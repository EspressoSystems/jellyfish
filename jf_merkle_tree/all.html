<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><title>List of all items in this crate</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-dd39b87e5fcfba68.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="jf_merkle_tree" data-themes="" data-resource-suffix="" data-rustdoc-version="1.80.0 (051478957 2024-07-21)" data-channel="1.80.0" data-search-js="search-d52510db62a78183.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-df360f571f6edeae.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod sys"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../jf_merkle_tree/index.html">jf_merkle_tree</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Aliases</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="append_only/struct.MerkleTree.html">append_only::MerkleTree</a></li><li><a href="examples/struct.Interval.html">examples::Interval</a></li><li><a href="hasher/struct.HasherDigestAlgorithm.html">hasher::HasherDigestAlgorithm</a></li><li><a href="hasher/struct.HasherNode.html">hasher::HasherNode</a></li><li><a href="light_weight/struct.LightWeightMerkleTree.html">light_weight::LightWeightMerkleTree</a></li><li><a href="namespaced_merkle_tree/struct.NMT.html">namespaced_merkle_tree::NMT</a></li><li><a href="prelude/struct.MerkleProof.html">prelude::MerkleProof</a></li><li><a href="prelude/struct.RescueHash.html">prelude::RescueHash</a></li><li><a href="prelude/struct.Sha3Digest.html">prelude::Sha3Digest</a></li><li><a href="prelude/struct.Sha3Node.html">prelude::Sha3Node</a></li><li><a href="universal_merkle_tree/struct.UniversalMerkleTree.html">universal_merkle_tree::UniversalMerkleTree</a></li></ul><h3 id="enums">Enums</h3><ul class="all-items"><li><a href="enum.LookupResult.html">LookupResult</a></li><li><a href="errors/enum.MerkleTreeError.html">errors::MerkleTreeError</a></li><li><a href="prelude/enum.MerkleNode.html">prelude::MerkleNode</a></li></ul><h3 id="traits">Traits</h3><ul class="all-items"><li><a href="trait.AppendableMerkleTreeScheme.html">AppendableMerkleTreeScheme</a></li><li><a href="trait.DigestAlgorithm.html">DigestAlgorithm</a></li><li><a href="trait.Element.html">Element</a></li><li><a href="trait.ForgetableMerkleTreeScheme.html">ForgetableMerkleTreeScheme</a></li><li><a href="trait.ForgetableUniversalMerkleTreeScheme.html">ForgetableUniversalMerkleTreeScheme</a></li><li><a href="trait.Index.html">Index</a></li><li><a href="trait.MerkleCommitment.html">MerkleCommitment</a></li><li><a href="trait.MerkleTreeScheme.html">MerkleTreeScheme</a></li><li><a href="trait.NodeValue.html">NodeValue</a></li><li><a href="trait.PersistentUniversalMerkleTreeScheme.html">PersistentUniversalMerkleTreeScheme</a></li><li><a href="trait.ToTraversalPath.html">ToTraversalPath</a></li><li><a href="trait.UniversalMerkleTreeScheme.html">UniversalMerkleTreeScheme</a></li><li><a href="hasher/trait.HasherDigest.html">hasher::HasherDigest</a></li><li><a href="namespaced_merkle_tree/trait.BindNamespace.html">namespaced_merkle_tree::BindNamespace</a></li><li><a href="namespaced_merkle_tree/trait.Namespace.html">namespaced_merkle_tree::Namespace</a></li><li><a href="namespaced_merkle_tree/trait.NamespaceProof.html">namespaced_merkle_tree::NamespaceProof</a></li><li><a href="namespaced_merkle_tree/trait.Namespaced.html">namespaced_merkle_tree::Namespaced</a></li><li><a href="namespaced_merkle_tree/trait.NamespacedMerkleTreeScheme.html">namespaced_merkle_tree::NamespacedMerkleTreeScheme</a></li></ul><h3 id="macros">Macros</h3><ul class="all-items"><li><a href="macro.impl_forgetable_merkle_tree_scheme.html">impl_forgetable_merkle_tree_scheme</a></li><li><a href="macro.impl_merkle_tree_scheme.html">impl_merkle_tree_scheme</a></li><li><a href="macro.impl_to_traversal_path_biguint.html">impl_to_traversal_path_biguint</a></li><li><a href="macro.impl_to_traversal_path_primitives.html">impl_to_traversal_path_primitives</a></li><li><a href="prelude/macro.impl_to_traversal_path_biguint.html">prelude::impl_to_traversal_path_biguint</a></li><li><a href="prelude/macro.impl_to_traversal_path_primitives.html">prelude::impl_to_traversal_path_primitives</a></li></ul><h3 id="types">Type Aliases</h3><ul class="all-items"><li><a href="examples/type.IntervalMerkleTree.html">examples::IntervalMerkleTree</a></li><li><a href="hasher/type.GenericHasherMerkleTree.html">hasher::GenericHasherMerkleTree</a></li><li><a href="hasher/type.HasherMerkleTree.html">hasher::HasherMerkleTree</a></li><li><a href="prelude/type.LightWeightSHA3MerkleTree.html">prelude::LightWeightSHA3MerkleTree</a></li><li><a href="prelude/type.MerklePath.html">prelude::MerklePath</a></li><li><a href="prelude/type.RescueLightWeightMerkleTree.html">prelude::RescueLightWeightMerkleTree</a></li><li><a href="prelude/type.RescueMerkleTree.html">prelude::RescueMerkleTree</a></li><li><a href="prelude/type.RescueSparseMerkleTree.html">prelude::RescueSparseMerkleTree</a></li><li><a href="prelude/type.SHA3MerkleTree.html">prelude::SHA3MerkleTree</a></li></ul></section></div></main></body></html>