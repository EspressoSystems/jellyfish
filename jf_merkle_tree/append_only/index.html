<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Implementation of a typical append only merkle tree"><title>jf_merkle_tree::append_only - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="jf_merkle_tree" data-themes="" data-resource-suffix="" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../../static.files/storage-82c7156e.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../jf_merkle_tree/index.html">jf_<wbr>merkle_<wbr>tree</a><span class="version">0.2.2</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module append_<wbr>only</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate jf_<wbr>merkle_<wbr>tree</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">jf_merkle_tree</a></div><h1>Module <span>append_only</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/jf_merkle_tree/append_only.rs.html#7-340">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Implementation of a typical append only merkle tree</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.MerkleTree.html" title="struct jf_merkle_tree::append_only::MerkleTree">Merkle<wbr>Tree</a></dt><dd>A standard append only Merkle tree implementation</dd></dl></section></div></main></body></html>