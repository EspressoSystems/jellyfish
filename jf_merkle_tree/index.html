<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Merkle Tree traits and implementations"><title>jf_merkle_tree - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="jf_merkle_tree" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.2 (25ef9e3d8 2024-04-09)" data-channel="1.77.2" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-4c98445ec4002617.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../jf_merkle_tree/index.html">jf_merkle_tree</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><ul class="block">
            <li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li></ul></section></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../jf_merkle_tree/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">jf_merkle_tree</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/jf_merkle_tree/lib.rs.html#7-449">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Merkle Tree traits and implementations</p>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.MerkleTreeError"><code>pub use crate::errors::<a class="enum" href="errors/enum.MerkleTreeError.html" title="enum jf_merkle_tree::errors::MerkleTreeError">MerkleTreeError</a>;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="append_only/index.html" title="mod jf_merkle_tree::append_only">append_only</a></div><div class="desc docblock-short">Implementation of a typical append only merkle tree</div></li><li><div class="item-name"><a class="mod" href="errors/index.html" title="mod jf_merkle_tree::errors">errors</a></div><div class="desc docblock-short">Error types</div></li><li><div class="item-name"><a class="mod" href="examples/index.html" title="mod jf_merkle_tree::examples">examples</a></div><div class="desc docblock-short">Provides sample instantiations of merkle tree.
E.g. Sparse merkle tree with BigUInt index.</div></li><li><div class="item-name"><a class="mod" href="hasher/index.html" title="mod jf_merkle_tree::hasher">hasher</a></div><div class="desc docblock-short">A convenience wrapper <a href="hasher/type.HasherMerkleTree.html" title="type jf_merkle_tree::hasher::HasherMerkleTree"><code>HasherMerkleTree</code></a> to instantiate <a href="append_only/struct.MerkleTree.html" title="struct jf_merkle_tree::append_only::MerkleTree"><code>MerkleTree</code></a> for any <a href="https://github.com/RustCrypto/hashes">RustCrypto-compatible</a> hash function.</div></li><li><div class="item-name"><a class="mod" href="light_weight/index.html" title="mod jf_merkle_tree::light_weight">light_weight</a></div><div class="desc docblock-short">A light weight merkle tree is an append only merkle tree who only keeps its
frontier – the right-most path.</div></li><li><div class="item-name"><a class="mod" href="macros/index.html" title="mod jf_merkle_tree::macros">macros</a></div><div class="desc docblock-short">Useful macros</div></li><li><div class="item-name"><a class="mod" href="namespaced_merkle_tree/index.html" title="mod jf_merkle_tree::namespaced_merkle_tree">namespaced_merkle_tree</a></div><div class="desc docblock-short">Implementation of a Namespaced Merkle Tree.</div></li><li><div class="item-name"><a class="mod" href="prelude/index.html" title="mod jf_merkle_tree::prelude">prelude</a></div><div class="desc docblock-short">Prelude. Also provides sample instantiations of merkle trees.</div></li><li><div class="item-name"><a class="mod" href="universal_merkle_tree/index.html" title="mod jf_merkle_tree::universal_merkle_tree">universal_merkle_tree</a></div><div class="desc docblock-short">Implementation of a typical Sparse Merkle Tree.</div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.impl_forgetable_merkle_tree_scheme.html" title="macro jf_merkle_tree::impl_forgetable_merkle_tree_scheme">impl_forgetable_merkle_tree_scheme</a></div><div class="desc docblock-short">Macro for generating a forgetable merkle tree implementation</div></li><li><div class="item-name"><a class="macro" href="macro.impl_merkle_tree_scheme.html" title="macro jf_merkle_tree::impl_merkle_tree_scheme">impl_merkle_tree_scheme</a></div><div class="desc docblock-short">Macro for generating a standard merkle tree implementation</div></li><li><div class="item-name"><a class="macro" href="macro.impl_to_traversal_path_biguint.html" title="macro jf_merkle_tree::impl_to_traversal_path_biguint">impl_to_traversal_path_biguint</a></div><div class="desc docblock-short">Macros for implementing ToTreversalPath for BigUint types</div></li><li><div class="item-name"><a class="macro" href="macro.impl_to_traversal_path_primitives.html" title="macro jf_merkle_tree::impl_to_traversal_path_primitives">impl_to_traversal_path_primitives</a></div><div class="desc docblock-short">Macros for implementing ToTreversalPath for primitive types</div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.LookupResult.html" title="enum jf_merkle_tree::LookupResult">LookupResult</a></div><div class="desc docblock-short">The result of querying at an index in the tree
Typically, F for element type, P for membership proof type and N for
non-membership proof type</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.AppendableMerkleTreeScheme.html" title="trait jf_merkle_tree::AppendableMerkleTreeScheme">AppendableMerkleTreeScheme</a></div><div class="desc docblock-short">Merkle tree that allows insertion at back. Abstracted as a commitment for
append-only vector.</div></li><li><div class="item-name"><a class="trait" href="trait.DigestAlgorithm.html" title="trait jf_merkle_tree::DigestAlgorithm">DigestAlgorithm</a></div><div class="desc docblock-short">Merkle tree hash function</div></li><li><div class="item-name"><a class="trait" href="trait.Element.html" title="trait jf_merkle_tree::Element">Element</a></div><div class="desc docblock-short">An element of a Merkle tree.</div></li><li><div class="item-name"><a class="trait" href="trait.ForgetableMerkleTreeScheme.html" title="trait jf_merkle_tree::ForgetableMerkleTreeScheme">ForgetableMerkleTreeScheme</a></div><div class="desc docblock-short">Merkle tree that allows forget/remember elements from the memory</div></li><li><div class="item-name"><a class="trait" href="trait.ForgetableUniversalMerkleTreeScheme.html" title="trait jf_merkle_tree::ForgetableUniversalMerkleTreeScheme">ForgetableUniversalMerkleTreeScheme</a></div><div class="desc docblock-short">Universal Merkle tree that allows forget/remember elements from the memory</div></li><li><div class="item-name"><a class="trait" href="trait.Index.html" title="trait jf_merkle_tree::Index">Index</a></div><div class="desc docblock-short">An index type of a leaf in a Merkle tree.</div></li><li><div class="item-name"><a class="trait" href="trait.MerkleCommitment.html" title="trait jf_merkle_tree::MerkleCommitment">MerkleCommitment</a></div><div class="desc docblock-short">Trait for a succinct merkle tree commitment</div></li><li><div class="item-name"><a class="trait" href="trait.MerkleTreeScheme.html" title="trait jf_merkle_tree::MerkleTreeScheme">MerkleTreeScheme</a></div><div class="desc docblock-short">Basic functionalities for a merkle tree implementation. Abstracted as an
accumulator for fixed-length array. Supports generate membership proof at a
given position and verify a membership proof.</div></li><li><div class="item-name"><a class="trait" href="trait.NodeValue.html" title="trait jf_merkle_tree::NodeValue">NodeValue</a></div><div class="desc docblock-short">An internal node value type in a Merkle tree.</div></li><li><div class="item-name"><a class="trait" href="trait.PersistentUniversalMerkleTreeScheme.html" title="trait jf_merkle_tree::PersistentUniversalMerkleTreeScheme">PersistentUniversalMerkleTreeScheme</a></div><div class="desc docblock-short">A universal merkle tree that allows non destructive updates.
A non destructive update doesn’t directly modify the existing content, it
creates a new copy about the update so that people could access both the old
version and the new.</div></li><li><div class="item-name"><a class="trait" href="trait.ToTraversalPath.html" title="trait jf_merkle_tree::ToTraversalPath">ToTraversalPath</a></div><div class="desc docblock-short">An trait for Merkle tree index type.</div></li><li><div class="item-name"><a class="trait" href="trait.UniversalMerkleTreeScheme.html" title="trait jf_merkle_tree::UniversalMerkleTreeScheme">UniversalMerkleTreeScheme</a></div><div class="desc docblock-short">A universal merkle tree is abstracted as a random-access array or a
key-value map. It allows manipulation at any given position, and has ability
to generate/verify a non-membership proof.</div></li></ul></section></div></main></body></html>