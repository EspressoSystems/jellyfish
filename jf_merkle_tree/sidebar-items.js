window.SIDEBAR_ITEMS = {"constant":["FAIL","SUCCESS"],"enum":["LookupResult"],"macro":["impl_forgetable_merkle_tree_scheme","impl_merkle_tree_scheme","impl_to_traversal_path_biguint","impl_to_traversal_path_primitives"],"mod":["append_only","errors","examples","hasher","light_weight","macros","prelude","universal_merkle_tree"],"trait":["AppendableMerkleTreeScheme","DigestAlgorithm","Element","ForgetableMerkleTreeScheme","ForgetableUniversalMerkleTreeScheme","Index","MerkleProof","MerkleTreeScheme","NodeValue","PersistentUniversalMerkleTreeScheme","ToTraversalPath","UniversalMerkleTreeScheme"]};