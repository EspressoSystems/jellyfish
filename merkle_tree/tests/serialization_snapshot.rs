//! Serialization snapshot tests for merkle tree structures used in Espresso
//! network.
//!
//! Tests serialization/deserialization of commitments and proofs to detect
//! breaking changes. In espresso-network we depend on compatibility of
//! commitments and serialization of proofs. If any of the tests in this file
//! fail the change in jellyfish would break the application.
//!
//! Run tests with `cargo test`, then review snapshots with `cargo insta review`
//! if new snapshots are added or tests fail. See `cargo insta --help` for more.
//!
//! For manual workflow (without `cargo insta`)
//! - reject: `INSTA_UPDATE=no cargo test`
//! - accept: `INSTA_UPDATE=always cargo test`.

use insta::assert_json_snapshot;
use jf_merkle_tree::{
    prelude::{LightWeightSHA3MerkleTree, Sha3Digest, Sha3Node},
    universal_merkle_tree::UniversalMerkleTree,
    AppendableMerkleTreeScheme, MerkleTreeScheme, UniversalMerkleTreeScheme,
};

// These aren't the exact trees used in epresso-network but are
// matching kind of trees and have matching arity
type BlockTree = LightWeightSHA3MerkleTree<u64>;
type FeeTree = UniversalMerkleTree<u64, Sha3Digest, u64, 256, Sha3Node>;
type RewardsTree = UniversalMerkleTree<u64, Sha3Digest, u64, 2, Sha3Node>;

#[test]
fn test_fee_merkle_tree_serialization() {
    let mut tree = FeeTree::new(8);

    let fee_data = vec![
        (1u64, 100u64),
        (2u64, 250u64),
        (3u64, 500u64),
        (4u64, 1000u64),
        (5u64, 2000u64),
    ];
    for (tx_id, fee) in fee_data {
        tree.update(tx_id, fee).unwrap();
    }

    let commitment = tree.commitment();
    assert_json_snapshot!("fee_tree_commitment", commitment);

    let mut proofs = Vec::new();
    for tx_id in 1u64..=5u64 {
        if let jf_merkle_tree::LookupResult::Ok(elem, proof) = tree.universal_lookup(tx_id) {
            proofs.push((tx_id, elem, proof));
        }
    }

    assert_json_snapshot!("fee_membership_proofs", proofs);
}

#[test]
fn test_block_merkle_tree_serialization() {
    let mut tree = BlockTree::new(8);

    let block_data = vec![
        0x1234567890abcdefu64,
        0xfedcba0987654321u64,
        0x1111222233334444u64,
        0x5555666677778888u64,
        0x9999aaaabbbbccccu64,
        0xddddeeeeffff0000u64,
    ];

    for block in block_data {
        tree.push(block).unwrap();
    }

    let commitment = tree.commitment();
    assert_json_snapshot!("block_tree_commitment", commitment);

    let mut proofs = Vec::new();
    for i in 0..tree.num_leaves() {
        if let jf_merkle_tree::LookupResult::Ok(elem, proof) = tree.lookup(i) {
            proofs.push((i, elem, proof));
        }
    }

    assert_json_snapshot!("block_membership_proofs", proofs);
}

#[test]
fn test_rewards_tree_serialization() {
    let mut tree = RewardsTree::new(8);

    let rewards = vec![
        (10u64, 1000u64),
        (25u64, 2500u64),
        (50u64, 5000u64),
        (100u64, 10000u64),
    ];

    for &(account, reward) in &rewards {
        tree.update(account, reward).unwrap();
    }

    let commitment = tree.commitment();
    assert_json_snapshot!("rewards_tree_commitment", commitment);

    let mut proofs = Vec::new();
    for &(account, _reward) in &rewards {
        if let jf_merkle_tree::LookupResult::Ok(elem, proof) = tree.universal_lookup(account) {
            proofs.push((account, elem, proof));
        }
    }

    assert_json_snapshot!("rewards_membership_proofs", proofs);
}

#[test]
fn test_rewards_tree_non_membership_proofs() {
    let mut tree = RewardsTree::new(8);

    let rewards = vec![(10u64, 1000u64), (50u64, 5000u64), (100u64, 10000u64)];

    for &(account, reward) in &rewards {
        tree.update(account, reward).unwrap();
    }

    let non_member_accounts = vec![5u64, 15u64, 75u64, 200u64];
    let mut non_membership_proofs = Vec::new();

    for &account in &non_member_accounts {
        if let jf_merkle_tree::LookupResult::NotFound(proof) = tree.universal_lookup(account) {
            non_membership_proofs.push((account, proof));
        }
    }

    assert_json_snapshot!("rewards_non_membership_proofs", non_membership_proofs);
}

#[test]
fn test_fee_tree_non_membership_proofs() {
    let mut tree = FeeTree::new(8);

    let fees = vec![(1u64, 100u64), (5u64, 250u64), (10u64, 500u64)];

    for &(tx_id, fee) in &fees {
        tree.update(tx_id, fee).unwrap();
    }

    let non_member_txs = vec![2u64, 3u64, 7u64, 15u64];
    let mut non_membership_proofs = Vec::new();

    for &tx_id in &non_member_txs {
        if let jf_merkle_tree::LookupResult::NotFound(proof) = tree.universal_lookup(tx_id) {
            non_membership_proofs.push((tx_id, proof));
        }
    }

    assert_json_snapshot!("fee_non_membership_proofs", non_membership_proofs);
}

#[test]
fn test_empty_tree_serialization() {
    let tree = BlockTree::new(8);
    let commitment = tree.commitment();

    assert_json_snapshot!("empty_tree_commitment", commitment);
}

#[test]
fn test_single_element_tree_serialization() {
    let mut tree = BlockTree::new(4);
    tree.push(42u64).unwrap();

    let commitment = tree.commitment();
    assert_json_snapshot!("single_element_tree_commitment", commitment);

    if let jf_merkle_tree::LookupResult::Ok(elem, proof) = tree.lookup(0u64) {
        assert_json_snapshot!("single_element_proof", (0u64, elem, proof));
    }
}

#[test]
fn test_full_tree_serialization() {
    let mut tree = BlockTree::new(3);

    for i in 0..8 {
        tree.push(i * 10).unwrap();
    }

    let commitment = tree.commitment();
    assert_json_snapshot!("full_tree_commitment", commitment);
}
