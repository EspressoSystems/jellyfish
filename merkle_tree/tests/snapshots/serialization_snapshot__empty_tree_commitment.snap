---
source: merkle_tree/tests/serialization_snapshot.rs
expression: commitment
---
MERKLE_COMM~AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAEA
